### PR TITLE
feat(miniprotocols): Implement tx submission client

### PR DESCRIPTION
**Pull request by scarmuega**

### CODE REVIEW

File: pallas-miniprotocols/src/txsubmission/README.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# TxSubmission
```
Comment (Quantumplation) [inline pallas-miniprotocols/src/txsubmission/README.md:1]: Rendered: https://github.com/txpipe/pallas/blob/ee7608a6393cd917a2799d47153d30f3a48049ab/pallas-miniprotocols/src/txsubmission/README.md

File: pallas-miniprotocols/src/txsubmission/server.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    pub fn recv_message(&mut self) -> Result<Message, Error> {
+        self.assert_agency_is_theirs()?;
+        let msg = self.1.recv_full_msg().map_err(Error::ChannelError)?;
+        self.assert_inbound_state(&msg)?;
+
+        Ok(msg)
+    }
+
+    pub fn wait_for_init(&mut self) -> Result<(), Error> {
+        if self.0 != State::Init {
```
Comment (Quantumplation) [inline pallas-miniprotocols/src/txsubmission/server.rs:96]: Was wondering if I should put some kind of loop here, in case we're waiting for a connection or something, but kept it very minimal;

If there's interest, I could make a `recv_init` that does the dumb thing, and a `wait_for_init` that loops until someone connects and sends the init message.

**Comment by Quantumplation** [comment]
clippy pls

File: examples/n2c-miniprotocols/src/main.rs [hunk 3]
```diff
@@ -66,7 +66,7 @@ fn main() {
 
     // we connect to the unix socket of the local node. Make sure you have the right
     // path for your environment
-    let bearer = Bearer::connect_unix("/tmp/node.socket").unwrap();
+    let bearer = Bearer::connect_tcp("/tmp/node.socket").unwrap();
```
Comment (scarmuega) [inline examples/n2c-miniprotocols/src/main.rs:69]: @Quantumplation why this change?

File: pallas-miniprotocols/src/txsubmission/server.rs (hunk 2 above)
Comment (scarmuega) [reply pallas-miniprotocols/src/txsubmission/server.rs:93]: well, the `recv_message` blocks until there's something to read from the channel, the "wait" is imposed to the consumer, even if they don't want it.

Ideally, every blocking call should provide an optional timeout, so that the control flow logic can escape the blocking call if desired. This is an improvement we should make at the multiplexer level, there should be a way to read from a channel with a timeout.

Comment (Quantumplation) [reply pallas-miniprotocols/src/txsubmission/server.rs:93]: Right but if someone hasn't even connected yet this errors immediately; or is there a construct at the multiplexer / channel level that lets me wait for a connection?

File: examples/n2c-miniprotocols/src/main.rs (hunk 3 above)
Comment (Quantumplation) [reply examples/n2c-miniprotocols/src/main.rs:69]: Because cargo test said the method didn't exist, and so was failing; I figured the method had been renamed but not updated in the docs

Comment (scarmuega) [reply examples/n2c-miniprotocols/src/main.rs:69]: oh, I think it's because that method is compiled conditionally only on OS that supports unix sockets.

File: pallas-miniprotocols/src/txsubmission/server.rs (hunk 2 above)
Comment (scarmuega) [reply pallas-miniprotocols/src/txsubmission/server.rs:93]: I see. I think that our server agent should be created as a result of a successful bearer connection (aka: tcp or unix socket connection).

I imagine the root process waiting for connections on a specific port, this would be a blocking call that is handled by the OS (something like `TcpListener::bind("127.0.0.1:80")`). Upon each connection, the process would create a new agent passing a channel connected to our new bearer (inbound socket stream). By this point in time, the calls to `recv_message` should not fail, but rather block waiting for data.

Summarizing, I think that the initial loop / block is out-of-scope for this crate. I say we create a new example crate (different PR) that shows how to connect our agent to a tcp listener. What do you think?

File: examples/n2c-miniprotocols/src/main.rs (hunk 3 above)
Comment (Quantumplation) [reply examples/n2c-miniprotocols/src/main.rs:69]: Should I revert it? Is there a way to mark the documentation/example conditionally in a similar way?

File: pallas-miniprotocols/src/txsubmission/server.rs (hunk 2 above)
Comment (Quantumplation) [reply pallas-miniprotocols/src/txsubmission/server.rs:93]: Makes sense to me; there's nothing in the type system that currently enforces this, though; i.e. I would expect something guiding me to "hey, you opened a connection to the relay, not the other way around, so you can't use this socket/channel to create a server"; see the integration test I added for example.

Is it worth adding a ClientChannel and ServerChannel type to distinguish this?

File: examples/n2c-miniprotocols/src/main.rs (hunk 3 above)
Comment (scarmuega) [reply examples/n2c-miniprotocols/src/main.rs]: I've added a conditional panic to make it explicit, let me know what you think.

**Comment by scarmuega** [comment]
@Quantumplation I think this is ready to merge. We can implement the improvements we discussed as independent PRs. Can you please give it a final review?

**Comment by Quantumplation** [comment]
@scarmuega LGTM!

**Review by Quantumplation — APPROVED** [review]
